Fix 6323: compress the bundle configMap

nunnatsa opened the pull request:
## Description of the change
Add the `--gzip-configmap=true` CLI parameter in order to compress the configMap, if it exeeds the max length.

## Motivation for the change
Fix #6323 

It is currently not possible to run bundles with files larger than 1MB (the configMap limit). In these cases, the existing partitioning mechanism is not enough, and the action fails.

OLM solved this issue by compressing the configMap content. This PR motivation is to implement compression mechanism in operator-sdk as well.

By adding the `--gzip-configmap=true` parameter, operator-sdk now will compress the content of the config map. The commad of the pod is also modified in this case. It is not possible to extract the files in the configMap itself. The command now extracts the files to `/var/tmp` which is writable, and then run the opm server form this directory.

Successfully manually tested on a local kind cluster, with:
```shell
$ build/operator-sdk run bundle --gzip-configmap=true quay.io/webcenter/elasticsearch-operator-bundle:v0.0.1 -n default

INFO[0009] Creating a File-Based Catalog of the bundle "quay.io/webcenter/elasticsearch-operator-bundle:v0.0.1" 
INFO[0011] Generated a valid File-Based Catalog         
INFO[0014] Created registry pod: quay-io-webcenter-elasticsearch-operator-bundle-v0-0-1 
INFO[0014] Created CatalogSource: elasticsearch-operator-catalog 
INFO[0014] Created Subscription: elasticsearch-operator-v0-0-1-sub 
INFO[0015] Approved InstallPlan install-z8jv8 for the Subscription: elasticsearch-operator-v0-0-1-sub 
INFO[0015] Waiting for ClusterServiceVersion "default/elasticsearch-operator.v0.0.1" to reach 'Succeeded' phase 
INFO[0015]   Waiting for ClusterServiceVersion "default/elasticsearch-operator.v0.0.1" to appear 
INFO[0028]   Found ClusterServiceVersion "default/elasticsearch-operator.v0.0.1" phase: Pending 
INFO[0030]   Found ClusterServiceVersion "default/elasticsearch-operator.v0.0.1" phase: Installing 
INFO[0051]   Found ClusterServiceVersion "default/elasticsearch-operator.v0.0.1" phase: Succeeded 
INFO[0051] OLM has successfully installed "elasticsearch-operator.v0.0.1" 
```

## Checklist

If the pull request includes user-facing changes, extra documentation is required:
- [x] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [x] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

everettraven reviewed (changes requested):
@nunnatsa Thanks a ton for taking a look at this and making this contribution! Overall this PR looks good, but there is a couple changes I'd like to see before I give this a thumbs up:
- For readability purposes I would prefer to see a single config map writer implementation that has configurable options. In the implementation we should try to generalize as much of the process as possible and only branch off when necessary for doing logic specific to an option.
- Use an init container to decompress the gzipped files before they are mounted to the regular container running `opm serve ...` - this helps us ensure that `gzip` is always available instead of assuming it is available within the index image.

I'll do another round of reviews once these changes are addressed :smile:

nunnatsa commented:
Hi @everettraven! Thanks for your comments.

I think that if I'll implement the gzip writer and the default writer on the same method, the result will be not readable nor maintainable. We'll need to use `if` for any other line. 

In the previous implementation, I used the "strategy" design pattern. I think it's great for our case, but I get your point about code duplication. So I now changed it to use the "template method" design pattern. I think it now gives us a good balance between readability and avoiding code duplication.

I also implemented your request about init container. 

Again, manually tested on kind cluster:

```shell
$ build/operator-sdk run bundle --gzip-configmap=true quay.io/webcenter/elasticsearch-operator-bundle:v0.0.1 -n default

INFO[0010] Creating a File-Based Catalog of the bundle "quay.io/webcenter/elasticsearch-operator-bundle:v0.0.1" 
INFO[0012] Generated a valid File-Based Catalog         
INFO[0015] Created registry pod: quay-io-webcenter-elasticsearch-operator-bundle-v0-0-1 
INFO[0015] Created CatalogSource: elasticsearch-operator-catalog 
INFO[0015] OperatorGroup "operator-sdk-og" created      
INFO[0015] Created Subscription: elasticsearch-operator-v0-0-1-sub 
INFO[0020] Approved InstallPlan install-w5kc7 for the Subscription: elasticsearch-operator-v0-0-1-sub 
INFO[0020] Waiting for ClusterServiceVersion "default/elasticsearch-operator.v0.0.1" to reach 'Succeeded' phase 
INFO[0020]   Waiting for ClusterServiceVersion "default/elasticsearch-operator.v0.0.1" to appear 
INFO[0031]   Found ClusterServiceVersion "default/elasticsearch-operator.v0.0.1" phase: Pending 
INFO[0033]   Found ClusterServiceVersion "default/elasticsearch-operator.v0.0.1" phase: InstallReady 
INFO[0034]   Found ClusterServiceVersion "default/elasticsearch-operator.v0.0.1" phase: Installing 
INFO[0054]   Found ClusterServiceVersion "default/elasticsearch-operator.v0.0.1" phase: Succeeded 
INFO[0054] OLM has successfully installed "elasticsearch-operator.v0.0.1" 
```

everettraven commented on internal/olm/operator/registry/fbcindex/configMapWriter.go:
```diff
@@ -0,0 +1,225 @@
+	return nil
+}
+
+type gzipCMWriter struct {
+	actualBuff   *bytes.Buffer
+	helperBuff   *bytes.Buffer
+	actualWriter *gzip.Writer
+	helperWriter *gzip.Writer
+	cmName       string
+	namespace    string
+}
+
+func newGZIPWriter(name, namespace string) *gzipCMWriter {
+	actualBuff := &bytes.Buffer{}
+	helperBuff := &bytes.Buffer{}
+
+	return &gzipCMWriter{
+		actualBuff:   actualBuff,
+		helperBuff:   helperBuff,
+		actualWriter: gzip.NewWriter(actualBuff),
+		helperWriter: gzip.NewWriter(helperBuff),
+		cmName:       name,
+		namespace:    namespace,
+	}
+}
```
I still think that having multiple config map writer implementations is unnecessary. I think we could get way with doing something like:
```suggestion
type configMapWriter struct {
	actualBuff   *bytes.Buffer
	helperBuff   *bytes.Buffer
	actualWriter io.Writer
	helperWriter io.Writer
	cmName       string
	namespace    string
	buffLimit       int
}

func newConfigMapWriter(name, namespace string, gzipped bool) *configMapWriter {
	actualBuff := &bytes.Buffer{}
	helperBuff := &bytes.Buffer{}

	cmWriter := &configMapWriter{
		actualBuff:   actualBuff,
		helperBuff:   helperBuff,
		actualWriter: bufio.NewWriter(actualBuff)
		helperWriter: bufio.NewWriter(helperBuff)
		cmName:       name,
		namespace:    namespace,
		buffLimit: maxConfigMapSize,
	}
	
	if gzipped {
	    cmWriter.actualWriter = gzip.NewWriter(actualBuff)
	    cmWriter.helperWriter = gzip.NewWriter(helperBuff)
	    buffLimit = maxGZIPLength
	}
	
	return cmWriter
}
```

This lets us use the exact same implementation for both while just adjusting the writer and buffer limit used when the gzip flag is passed.

nunnatsa replied:
I still think it's a very good implementation. I think the suggested solution above misses many of the differences between the two writers. The design pattern in use here is well known and it's a solid solution for similar challenges. 

There are several advantage for this approach. First, it detaches the partitioning algorithm from the actual writing, kipping most of the flow common and relatively easy to read. The actual write to the configMap, and all the rest of the specifications of actual chosen configMap type, are implemented elsewhere. 

The main flow does not have branches every other line, and each writer can be very tight and accurate.

Bottom line - as I see it - this approach is easier to implement, read and maintain, and follows well known design pattern.

Another advantage, that I don't think will be related here, is if we'll need to add a third type of configMap (e.g. json based), we'll only need to implement a new writer.

everettraven replied:
> I still think it's a very good implementation. I think the suggested solution above misses many of the differences between the two writers. The design pattern in use here is well known and it's a solid solution for similar challenges.

Is there really a major difference between the default and gzip ConfigMap writer? From what I can tell it would just be the difference of using a `gzip.NewWriter` versus a `bufio.NewWriter` and all the underlying functions implement the same logic.

> There are several advantage for this approach. First, it detaches the partitioning algorithm from the actual writing, kipping most of the flow common and relatively easy to read. The actual write to the configMap, and all the rest of the specifications of actual chosen configMap type, are implemented elsewhere.
>
> The main flow does not have branches every other line, and each writer can be very tight and accurate.

I don't think the solution I suggested changes this though. It would still have a the same functional breakdown, but we would only be changing what implementation of the `io.Writer` interface we are using when we write the specified data.


> Bottom line - as I see it - this approach is easier to implement, read and maintain, and follows a well known design pattern.

I'm not saying that this is a bad or unreadable design pattern, I'm just questioning whether or not we actually need to have 2 separate implementations that implement the `configMapWriter` interface when it seems to me that the difference between the two becomes using a `gzip.NewWriter()` function vs a `bufio.NewWriter()` function to initialize the writer type. I think all the other `gzipConfigMapWriter` functions are general enough that we just swap out the `.actualWriter` and `.helperWriter` values to be whatever `io.Writer` implementation we want.

> Another advantage, that I don't think will be related here, is if we'll need to add a third type of configMap (e.g. json based configMap), we'll only need to implement a new writer.

We could still keep the interface around and have this ability (and having the interface makes it easier to test).

--- 

Overall I don't disagree with your points or the design of the implementation. My questioning is strictly around whether or not we *need* 2 separate implementations of the `configMapWriter` interface.

everettraven replied:
So looking at this more it does look like there is the difference of *where* the data is written to in the ConfigMap (`ConfigMap.Data` vs `ConfigMap.BinaryData`) but I still think that could be generalized by an internal function used to write the data to the right ConfigMap field.

joelanford replied:
Could we _always_ write to the binary data field?

nunnatsa replied:
> Could we _always_ write to the binary data field?

I don't think so. textual data should be in the `Data` field.

nunnatsa replied:
Thanks again @everettraven for you patience. I'm not sure I get your point. What wrong with multiple implementations for an interface? this is the whole idea. We **want** to have multiple implementations. I think we gain many advantages going this path. As I said, this is not my invention - this programming challenge is not new, nor its solutions. 

Specifically, changing the implementation of the default writer to work with two internal writer just to not having multiple implementation of the interface, is not a good reason, I think, because we will write every segment of the data twice. We don't need this for the default implementation, since we can know the length without writing the next segment. This is not the case for the compression implementation, where we must write the next segment to check the new length.

So the solution will work, but it will be less efficient.

everettraven commented on internal/olm/operator/config.go:
```diff
@@ -62,6 +63,8 @@ func (c *Configuration) BindFlags(fs *pflag.FlagSet) {
 			"This value does not override the operator's service account")
 	fs.DurationVar(&c.Timeout, "timeout", 2*time.Minute,
 		"Duration to wait for the command to complete before failing")
+	fs.BoolVar(&c.GzipCM, "gzip-configmap", false, `If 'true', the configmap data will be compressed. `+
```
Nit: I feel like `--gzip-configmap` is pretty long to type out I would prefer `--gzip`

everettraven commented on internal/olm/operator/config.go:
```diff
@@ -62,6 +63,8 @@ func (c *Configuration) BindFlags(fs *pflag.FlagSet) {
 	fs.DurationVar(&c.Timeout, "timeout", 2*time.Minute,
 		"Duration to wait for the command to complete before failing")
+	fs.BoolVar(&c.GzipCM, "gzip-configmap", false, `If 'true', the configmap data will be compressed. `+
+		`Use if in case of getting the like "... ConfigMap ... is invalid: []: Too long: must have at most 1048576 bytes"`)
```
Nit: some minor wording tweaks
```suggestion
		`Useful when getting errors with a message similar to "... ConfigMap ... is invalid: []: Too long: must have at most 1048576 bytes"`)
```

everettraven reviewed (commented):
@nunnatsa Thanks for bearing with my delay in getting a review in. I took another look at the `configMapWriter` implementation and left some suggestions. I took a look at the rest of the changes and they seem reasonable but I was having some trouble with the GitHub diff formatting. I have to drop off for the day but I will circle back around and do another round of review on Monday.

Thanks again for your patience!

everettraven commented on internal/olm/operator/registry/fbcindex/fbc_registry_pod.go:
```diff
@@ -233,8 +237,11 @@ func (f *FBCRegistryPod) podForBundleRegistry(cs *v1alpha1.CatalogSource) (*core
 
-	volumes := []corev1.Volume{}
-	volumeMounts := []corev1.VolumeMount{}
+	var (
+		volumes            []corev1.Volume
+		sharedVolumeMounts []corev1.VolumeMount
+		gzipVolumeMount    []corev1.VolumeMount
+	)
```
Nit: I *personally* prefer to do something like:
```go
volumes := []corev1.Volume{}
sharedVolumeMounts := []corev1.VolumeMount{}
gzipVolumeMounts := []corev1.VolumeMount{}
```

This shouldn't block the merge though if everything else checks out.

nunnatsa replied:
I don't have a strong opinion about it, but according to the golang wiki, this format is prefered: https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices.

everettraven commented on internal/olm/operator/registry/fbcindex/fbc_registry_pod.go:
```diff
@@ -254,13 +261,44 @@ func (f *FBCRegistryPod) podForBundleRegistry(cs *v1alpha1.CatalogSource) (*core
+			})
+		}
+
+		postfix := cm.Name + "-volume"
```
Does this need the `cm.Name`? This seems unnecessary if we are also explicitly adding it on [line 279](https://github.com/operator-framework/operator-sdk/pull/6408/files?diff=split&w=0#diff-aba7d9ab4132a0f10f2e4be48a47e16124327c89a22d2de694cba5b5c22ae324R279)

```suggestion
		postfix := "-volume"
```

everettraven commented on internal/olm/operator/registry/fbcindex/fbc_registry_pod.go:
```diff
@@ -254,13 +261,44 @@ func (f *FBCRegistryPod) podForBundleRegistry(cs *v1alpha1.CatalogSource) (*core
-		})
+		}
+
+		if !f.cfg.GzipCM {
+			vm.SubPath = cm.Name
+		}
```
The `VolumeMount.SubPath` is already `cm.Name`. I would remove this if block all together

everettraven commented on internal/olm/operator/registry/fbcindex/fbc_registry_pod.go:
```diff
@@ -233,8 +237,11 @@ func (f *FBCRegistryPod) podForBundleRegistry(cs *v1alpha1.CatalogSource) (*core
+	var (
+		volumes            []corev1.Volume
+		sharedVolumeMounts []corev1.VolumeMount
+		gzipVolumeMount    []corev1.VolumeMount
```
Is there a specific reason we need to have a separate variable for this? Can we just add the gzip volume mount to the sharedVolumeMounts if the gzip flag is enabled?

nunnatsa replied:
For the regular usecasem these volumeMountes point to the configMaps, while in the gzip usecase, they point to the uncompressed volumes.

The `gzipVolumeMount` slice only used in the gzip usecase, and its volumeMounts point to the configMaps. 

So in the gzip usecase, we want `sharedVolumeMountslice` slice both in t init container and in the registry container, and we want the `gzipVolumeMount` only in the init container.

everettraven commented on internal/olm/operator/registry/fbcindex/fbc_registry_pod.go:
```diff
@@ -254,13 +261,44 @@ func (f *FBCRegistryPod) podForBundleRegistry(cs *v1alpha1.CatalogSource) (*core
+		}
 	}
 
+	containerVolumeMount := make([]corev1.VolumeMount, len(sharedVolumeMounts))
+	copy(containerVolumeMount, sharedVolumeMounts)
```
Why do we need to do this copy? Can we just pass sharedVolumeMounts directly to wherever we are passing containerVolumeMount?

nunnatsa replied:
We don't. it's a leftover from a previous implementation. Removed.

everettraven reviewed (commented):
Did another round of review to run through some things I didn't get a chance to look at in the last one

nunnatsa commented:
> Did another round of review to run through some things I didn't get a chance to look at in the last one

Thanks @everettraven. Fixed most of it, answered to the rest.

everettraven commented on changelog/fragments/support_configmap_compression.yaml:
```diff
@@ -0,0 +1,19 @@
+# release notes and/or the migration guide
+entries:
+  - description: >
+      Allow compress the bundle content. Added a new cli flag `--gzip-configmap=true` to the `operator-sdk run bundle`.
```
Flag name changed in the implementation so we should make sure we change it in the changelog as well:
```suggestion
      Allow compress the bundle content. Added a new boolean cli flag `--gzip` to the `operator-sdk run bundle`.
```

everettraven commented on internal/olm/operator/registry/fbcindex/fbc_registry_pod.go:
```diff
@@ -315,25 +346,51 @@ func (f *FBCRegistryPod) podForBundleRegistry(cs *v1alpha1.CatalogSource) (*core
-const fbcCmdTemplate = `opm serve {{ .FBCIndexRootDir}} -p {{ .GRPCPort }}`
+func (f *FBCRegistryPod) addGZIPInitContainer(containerVolumeMount []corev1.VolumeMount, gzipVolumeMount []corev1.VolumeMount) {
+	initContainerVolumeMount := append(containerVolumeMount, gzipVolumeMount...)
+	f.pod.Spec.InitContainers = []corev1.Container{
```
Nit: Is the init containers slice already initialized before this? Only concern I have here being that if we added init containers somewhere else in the future this would overwrite them all to just be the gzip init container rather than appending it.

everettraven reviewed (commented):
Overall looks good to me and should be okay to merge. *Personally* I would still like to see a single `configMapWriter` interface implementation but I don't think that should block this PR if others think it looks good.

I would definitely like to get another set of eyes on this prior to merging though.

grokspawn commented:
/retest

joelanford commented:
Perhaps a dumb question, but if I'm wondering others may be as well: Why does this need to be a flag? Why not just _always_ gzip the config map contents?

acornett21 commented:
> Perhaps a dumb question, but if I'm wondering others may be as well: Why does this need to be a flag? Why not just _always_ gzip the config map contents?

@joelanford I'm also wondering the same thing. It's better usability for users to not have to re-run if they hit this issue.

nunnatsa commented:
> > Perhaps a dumb question, but if I'm wondering others may be as well: Why does this need to be a flag? Why not just _always_ gzip the config map contents?
> 
> @joelanford I'm also wondering the same thing. It's better usability for users to not have to re-run if they hit this issue.

I'm looking into it. meantime it looks good. Checking the upgrade usecase, when the previous version is not compressed.

nunnatsa commented:
/retest

grokspawn commented:
/ok-to-test

theishshah commented:
Going to take a look at these scorecard failures to see if there is a way we can help get this unblocked

nunnatsa commented:
updated the  digests of the scorecard-storage and scorecard-untar images, to unblock the scorecard tests.

nunnatsa commented:
/retest

nunnatsa commented:
@theishshah, @grokspawn - I think I know why the scorecard tests are failing, see #6428 and #6429

I fixed the image digest here in this PR too. Could you please run the test again?

nunnatsa commented:
@everettraven - could you please run the tests again?

nunnatsa commented on internal/cmd/operator-sdk/scorecard/cmd.go:
```diff
@@ -94,11 +94,11 @@ If the argument holds an image tag, it must be present remotely.`,
 	// If you need to make changes to these images, please ensure that you always use the digests.
 	scorecardCmd.Flags().StringVarP(&c.storageImage, "storage-image", "b",
-		"quay.io/operator-framework/scorecard-storage@sha256:f7bd62664a0b91034acb977a8bb4ebb76bc98a6e8bdb943eb84c8e364828f056",
+		"quay.io/operator-framework/scorecard-storage@sha256:5f9640f6eb6a6976676f2936b9eb4cd7170c5eebbc7536cc2891ec6cba74f0dd",
```
Dropped.

nunnatsa commented:
Rebased.

nunnatsa commented:
@grokspawn ^^^

everettraven commented on internal/olm/operator/registry/fbcindex/fbc_registry_pod_test.go:
```diff
@@ -164,53 +142,91 @@ var _ = Describe("FBCRegistryPod", func() {
+				expectedYaml := strings.TrimPrefix(strings.TrimSpace(largeYaml), "---\n")
+				expectedYaml = regexp.MustCompile(`\n\n+`).ReplaceAllString(expectedYaml, "\n")
+
+				// debug
+				length := float64(len(expectedYaml))
+				GinkgoWriter.Printf("expected yaml created; expected yaml length: %.0f bytes; %.2fkb; %.2fmb\n", length, length/1024, length/1024/1024)
+				// debug-end
```
We should probably clean up this debugging logic

everettraven commented on internal/olm/operator/registry/fbcindex/fbc_registry_pod_test.go:
```diff
@@ -164,53 +142,91 @@ var _ = Describe("FBCRegistryPod", func() {
+				cms, err := rp.partitionedConfigMaps()
+				Expect(err).ShouldNot(HaveOccurred())
+
+				// debug
+				length = float64(len(cms[0].BinaryData["extraFBC"])) + float64(len(cms[1].BinaryData["extraFBC"]))
+				GinkgoWriter.Printf("CM was created; expected yaml length: %.0f bytes; %.2fkb; %.2fmb\n", length, length/1024, length/1024/1024)
+				// debug-end
```
ditto

nunnatsa replied:
oops...
thanks!

nunnatsa commented:
thanks @everettraven - fixed your comments.

nunnatsa commented:
/retest

nunnatsa commented:
@grokspawn @everettraven @joelanford if this is ready, can we merge it?

everettraven commented:
> if this is ready, can we merge it?

@nunnatsa I'm adding this to my todo list to circle back around to tomorrow

everettraven commented:
@nunnatsa Once the branch conflict is resolved and tests have run I'll work on getting this merged (some CI tests are likely to fail but I'll need to verify they are failing for the expected reason)

nunnatsa commented:
Thanks @everettraven - now we have conflicts.

everettraven reviewed (approved):
/lgtm

everettraven commented:
Overriding the failing checks as they are the same issues being track in #6434 and #6433 

/override docs
/override e2e-molecule
